File: src/lidar_lite_ros/lidar_lite_ros/lidar_lite_node.py
#!/usr/bin/env python3
"""
ROS2 Node for Garmin Lidar Lite v3HP.
Publishes range data to `lidar/range`.
"""
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Range
from collections import deque
from lidar_lite_ros.lidar_lite_lib import LidarLite, DEFAULT_ADDRESS

class LPF:
    def __init__(self, alpha=0.9, max_change=0.5):
        """
        Low-pass filter with simple outlier rejection.
        
        Args:
            alpha: Smoothing factor (0.0 = no smoothing, 1.0 = no update)
            max_change: Maximum allowed change between readings (meters)
        """
        self.alpha = alpha
        self.max_change = max_change
        self.old_val = None
        
    def filter(self, new_val):
        if self.old_val is None:
            self.old_val = new_val
            return self.old_val
        
        # Check for outlier
        change = abs(new_val - self.old_val)
        if change > self.max_change:
            # Outlier detected - reject and keep old value
            return self.old_val
        
        # Normal LPF update
        self.old_val = self.alpha * self.old_val + (1 - self.alpha) * new_val
        return self.old_val
    
    def reset(self):
        """Reset filter state"""
        self.old_val = None

class LidarLiteNode(Node):
    def __init__(self):
        super().__init__('lidar_lite_node')
        
        # I2C Interface Parameters
        self.declare_parameter('i2c_interface.bus', 1)
        self.declare_parameter('i2c_interface.address', DEFAULT_ADDRESS)

        # ROS Parameters
        self.declare_parameter('frequency', 100.0) # Hz
        self.declare_parameter('frame_id', 'lidar_link')

        # Sensor Configuration Parameters
        self.declare_parameter('sensor_config.min_range', 0.05)
        self.declare_parameter('sensor_config.max_range', 40.0)
        self.declare_parameter('sensor_config.preset', 'balanced')
        self.declare_parameter('sensor_config.sig_count_val', -1)
        self.declare_parameter('sensor_config.acq_config_reg', -1)
        self.declare_parameter('sensor_config.threshold_bypass', -1)
        self.declare_parameter('sensor_config.ref_count_val', -1)
        
        self.bus_id = self.get_parameter('i2c_interface.bus').value
        self.address = self.get_parameter('i2c_interface.address').value

        self.freq = self.get_parameter('frequency').value
        self.frame_id = self.get_parameter('frame_id').value
        
        self.min_range = self.get_parameter('sensor_config.min_range').value
        self.max_range = self.get_parameter('sensor_config.max_range').value
        
        # Determine Final Settings
        preset_name = self.get_parameter('sensor_config.preset').value
        if preset_name not in LidarLite.PRESETS:
            self.get_logger().warn(f"Unknown preset '{preset_name}', falling back to 'balanced'")
            preset_name = 'balanced'
            
        settings = LidarLite.PRESETS[preset_name].copy()
        
        # Check for individual overrides
        for key in settings.keys():
            val = self.get_parameter(f'sensor_config.{key}').value
            if val != -1:
                settings[key] = val
                self.get_logger().info(f"Overriding {key} from preset with: {val}")

        # Hardware
        self.lidar = LidarLite(bus_id=self.bus_id, address=self.address)
        if self.lidar.connected:
            self.get_logger().info(f'✅ Lidar Lite init OK on bus {self.bus_id} at address 0x{self.address:02X}')
        else:
            self.get_logger().fatal(f'❌ Lidar Lite init FAILED on bus {self.bus_id} at address 0x{self.address:02X}')
            self._ok = False
            return
        
        if self.lidar.configure(settings):
            self.get_logger().info(f'✅ Lidar Lite configured with preset {preset_name}')
        else:
            self.get_logger().warn('❌ Failed to apply some Lidar Lite settings.')
            
        # Publisher
        self.pub_range = self.create_publisher(Range, "lidar/range", 10)
        
        # Timer
        period = 1.0 / self.freq
        self.timer = self.create_timer(period, self.tick)

        # Buffer for median filtering
        self.reading_buffer = deque(maxlen=5)  # Keep last 5 readings
        self.lpf = LPF(alpha=0.8)
        
        self.get_logger().info(f'Lidar Lite Node started. Bus: {self.bus_id}, Freq: {self.freq}Hz')

    def tick(self):
        dist = self.lidar.read_distance()
        
        if dist is None:
            self.get_logger().warn(f"Lidar Distance is None", throttle_duration_sec=1.0)
            return
        
        filtered_dist = self.lpf.filter(dist)
        
        msg = Range()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self.frame_id
        msg.radiation_type = Range.INFRARED
        msg.field_of_view = 0.01
        msg.min_range = self.min_range
        msg.max_range = self.max_range
        msg.range = filtered_dist
                
        self.pub_range.publish(msg)
      

    def destroy_node(self):
        if hasattr(self, 'lidar'):
            self.lidar.close()            
        if self.timer:
            self.timer.cancel()
        super().destroy_node()

def main(args=None):
    rclpy.init(args=args)
    node = LidarLiteNode()

    if not getattr(node, '_ok', True):
        rclpy.shutdown()
        return

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/lidar_lite_ros/src/lidar_lite_node.cpp
#include <memory>
#include <string>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/range.hpp"
#include "lidar_lite_ros/lidar_lite.hpp"

using namespace std::chrono_literals;

namespace lidar_lite_ros {

class LidarLiteNode : public rclcpp::Node {
public:
    LidarLiteNode() : Node("lidar_lite_node") {
        // Declare ROS2 parameters
        this->declare_parameter("i2c_bus", 1);
        this->declare_parameter("frequency", 100.0);
        this->declare_parameter("frame_id", "lidar_link");
        this->declare_parameter("min_range", 0.05);
        this->declare_parameter("max_range", 40.0);
        this->declare_parameter("preset", "balanced");
        
        // Individual overrides
        this->declare_parameter("sig_count_val", -1);
        this->declare_parameter("acq_config_reg", -1);
        this->declare_parameter("threshold_bypass", -1);
        this->declare_parameter("ref_count_val", -1);

        // Load parameters
        int bus_id = this->get_parameter("i2c_bus").as_int();
        double freq = this->get_parameter("frequency").as_double();
        frame_id_ = this->get_parameter("frame_id").as_string();
        min_range_ = this->get_parameter("min_range").as_double();
        max_range_ = this->get_parameter("max_range").as_double();
        std::string preset_name = this->get_parameter("preset").as_string();

        // Determine final settings
        LidarSettings settings;
        auto it = LidarLite::PRESETS.find(preset_name);
        if (it != LidarLite::PRESETS.end()) {
            settings = it->second;
            RCLCPP_INFO(this->get_logger(), "Using Lidar Lite preset: '%s'", preset_name.c_str());
        } else {
            RCLCPP_WARN(this->get_logger(), "Unknown preset '%s', falling back to 'balanced'", preset_name.c_str());
            settings = LidarLite::PRESETS.at("balanced");
        }

        // Apply overrides
        int val;
        val = this->get_parameter("sig_count_val").as_int();
        if (val != -1) settings.sig_count_val = static_cast<uint8_t>(val);
        
        val = this->get_parameter("acq_config_reg").as_int();
        if (val != -1) settings.acq_config_reg = static_cast<uint8_t>(val);
        
        val = this->get_parameter("threshold_bypass").as_int();
        if (val != -1) settings.threshold_bypass = static_cast<uint8_t>(val);
        
        val = this->get_parameter("ref_count_val").as_int();
        if (val != -1) settings.ref_count_val = static_cast<uint8_t>(val);

        // Initialize Hardware
        lidar_ = std::make_unique<LidarLite>(bus_id);
        if (!lidar_->isConnected()) {
            RCLCPP_ERROR(this->get_logger(), "Lidar Lite not detected on bus %d!", bus_id);
        } else {
            if (lidar_->configure(settings)) {
                RCLCPP_INFO(this->get_logger(), "Lidar Lite configured successfully");
            } else {
                RCLCPP_WARN(this->get_logger(), "Failed to apply some Lidar Lite settings");
            }
        }

        // Publisher
        pub_range_ = this->create_publisher<sensor_msgs::msg::Range>("lidar/range", 10);

        // Timer
        auto period = std::chrono::duration<double>(1.0 / freq);
        timer_ = this->create_wall_timer(period, std::bind(&LidarLiteNode::timer_callback, this));

        RCLCPP_INFO(this->get_logger(), "Lidar Lite Node started. Bus: %d, Freq: %.1fHz", bus_id, freq);
    }

private:
    void timer_callback() {
        auto dist = lidar_->readDistance();
        if (dist.has_value()) {
            auto msg = sensor_msgs::msg::Range();
            msg.header.stamp = this->now();
            msg.header.frame_id = frame_id_;
            msg.radiation_type = sensor_msgs::msg::Range::INFRARED;
            msg.field_of_view = 0.01f; // Approx 8mrad ~ 0.5 deg
            msg.min_range = static_cast<float>(min_range_);
            msg.max_range = static_cast<float>(max_range_);
            msg.range = dist.value();

            pub_range_->publish(msg);
        }
    }

    std::unique_ptr<LidarLite> lidar_;
    rclcpp::Publisher<sensor_msgs::msg::Range>::SharedPtr pub_range_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::string frame_id_;
    double min_range_;
    double max_range_;
};

} // namespace lidar_lite_ros

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<lidar_lite_ros::LidarLiteNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
